Downstream DNS Server Cookies a la RFC7873 and RFC9018

wtoorop opened the pull request:
Create server cookies for clients that send client cookies.
This needs to be explicitly turned on in the config file with:
```
	answer-cookie: yes
```    

A `cookie-secret:` can be configured for anycast setups. Without one, a random cookie secret is generated.

This PR also adds an access control list policy `allow_cookie` that allows queries with either a valid cookie or over a stateful transport.

- [ ] TODO: Rolling cookie-secrets

This PR replaces #322

wtoorop committed:
extended_error_encode() for extended errors

wtoorop committed:
Downstream DNS Cookies a la RFC7873 and RFC9018

Create server cookies for clients that send client cookies.
Needs to be turned on in the config file with:

	answer-cookie: yes

A cookie-secret can be configured for anycast setups.
Also adds an access control list that will allow queries with
either a valid cookie or over a stateful transport.

wtoorop committed:
The generated lexer and parser sources for configuring cookies

wtoorop committed:
Merge branch 'master' into devel/merge-master-into-downstream-cookies

gthess committed:
- For #762: fix compiler C90 warning.

gthess committed:
- For #762: More generic integration for siphash.c

gthess committed:
- For #762: relocate edns_opt_list_append_keepalive.

gthess committed:
- For #762: please doxygen.

gthess committed:
- For #762: relocate RFC 1982 serial number arithmetic functions to their own
  file in util/rfc_1982.[ch].

gthess committed:
- For #762: Ignore util/siphash.c for the 01-doc test.

gthess committed:
- For #762: remove relocated code.

gthess committed:
- For #762: relocate EDNS cookie code to util/edns and introduce unit
  tests.

gthess committed:
- For #762: Formatting.

gthess committed:
- For #762: Introduce rpl testing for DNS Cookies.

gthess committed:
- For #762: Cleaner manpage text and uniform use of the term DNS
  Cookies.

gthess committed:
- For #762: remove uneeded include.

gthess committed:
Merge branch 'master' into features/downstream-cookies

gthess committed:
- For #762: annotate case statement fallthrough for gcc.

gthess committed:
- For #762: Remove re-introduced files from merge (configlexer.c, configparser.c, configparser.h).

gthess committed:
- For #762: Silence maybe-uninitialized compiler warning.

gthess committed:
- For #762: Interaction between DNS Cookies and source IP ratelimiting
  by allowing Cookies to bypass the ratelimit, but still allowing
  ratelimit to valid DNS Cookie clients via the new
  ip-ratelimit-cookie option.

gthess committed:
- For #762: Introduce stat counters for downstream DNS Cookies per
  thread and total: num.queries_cookie_valid, num.queries_cookie_client,
  num.queries.cookie_invalid.

gthess commented:
@wcawijngaards, with the recent changes for refactoring, unit/rpl/tdir tests, cookie/ip_ratelimiting interaction and new stat counters for queries with cookies, I believe this is ready for merging after your final review :)

wcawijngaards committed:
- Fix possibly unaligned memory access.

wcawijngaards committed:
- Fix possibly unaligned memory access.

wcawijngaards committed:
- Fix out of bounds read in parse_edns_options_from_query, it would read
  8 bytes after a client option of length 8, and then ignore them to
  recreate a 24 byte response. The fixup does not read out of bounds,
  and puts zeroes in the buffer at that point, that then are ignored.

wcawijngaards reviewed (approved):
Reviewed the code and committed couple of fixes. Nice to have the additional feature.
